### ISSUE TITLE
registry:2 corrupts uploaded layers (filesystem layer verification failed) when using 3rd-party S3

### ISSUE THREAD

**Issue by zllovesuki** [docker/distribution#2205]
Edit 2: It looks like an issue of handling multipart offset on minio's part. The developers are actively trying to resolve the issue, and they proposed a fix (pending validation).

---

Edit 1: It is possible that 63468ef4a85fabb756e799c2534f0df2f3c8167c introduces this regression

---

Numerous similar issues: #2168, docker/docker#21290, etc

Running on Kubernetes cluster:
```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"clean", BuildDate:"2016-12-14T00:57:05Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1+coreos.0", GitCommit:"cc65f5321f9230bf9a3fa171155c1213d6e3480e", GitTreeState:"clean", BuildDate:"2016-12-14T04:08:28Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

Registry deployment file:
```json
{
    "apiVersion": "extensions/v1beta1",
    "kind": "Deployment",
    "metadata": {
        "name": "kube-registry",
        "namespace": "kube-system"
    },
    "spec": {
        "replicas": 3,
        "selector": {
            "matchLabels": {
                "app": "kube-registry"
            }
        },
        "template": {
            "metadata": {
                "labels": {
                    "app": "kube-registry"
                }
            },
            "spec": {
                "volumes": [{
                    "name": "auth-signing",
                    "secret": {
                        "secretName": "auth-signing"
                    }
                }],
                "containers": [{
                    "name": "sysctl-buddy",
                    "image": "alpine:latest",
                    "command": [
                        "/bin/sh",
                        "-c",
                        "while true; do\n  sysctl -w net.core.somaxconn=32768 > /dev/null 2>&1\n  sleep 10\ndone\n"
                    ],
                    "imagePullPolicy": "IfNotPresent",
                    "securityContext": {
                        "privileged": true
                    }
                }, {
                    "name": "registry",
                    "image": "registry:2",
                    "resources": {
                        "requests": {
                            "cpu": "200m",
                            "memory": "1024Mi"
                        },
                        "limits": {
                            "cpu": "2",
                            "memory": "2048Mi"
                        }
                    },
                    "env": [{
                            "name": "REGISTRY_LOG_LEVEL",
                            "value": "warn"
                        },
                        {
                            "name": "REGISTRY_HTTP_ADDR",
                            "value": ":5000"
                        },
                        {
                            "name": "REGISTRY_HTTP_HOST",
                            "value": "[redacted]"
                        },
                        {
                            "name": "REGISTRY_HTTP_SECRET",
                            "value": "[redacted]"
                        },
                        {
                            "name": "REGISTRY_STORAGE",
                            "value": "s3"
                        },
                        {
                            "name": "REGISTRY_STORAGE_DELETE_ENABLED",
                            "value": "true"
                        },
                        {
                            "name": "REGISTRY_STORAGE_S3_REGION",
                            "value": "us-east-1"
                        },
                        {
                            "name": "REGISTRY_STORAGE_S3_REGIONENDPOINT",
                            "value": "[redacted]"
                        },
                        {
                            "name": "REGISTRY_STORAGE_S3_BUCKET",
                            "value": "images"
                        },
                        {
                            "name": "REGISTRY_STORAGE_S3_ACCESSKEY",
                            "value": "docker"
                        },
                        {
                            "name": "REGISTRY_STORAGE_S3_SECRETKEY",
                            "value": "[redacted]"
                        },
                        {
                            "name": "REGISTRY_STORAGE_S3_ENCRYPT",
                            "value": "false"
                        },
                        {
                            "name": "REGISTRY_STORAGE_S3_SECURE",
                            "value": "false"
                        },
                        {
                            "name": "REGISTRY_STORAGE_S3_CHUNKSIZE",
                            "value": "20971520"
                        },
                        {
                            "name": "REGISTRY_STORAGE_CACHE_BLOBDESCRIPTOR",
                            "value": "redis"
                        },
                        {
                            "name": "REGISTRY_REDIS_ADDR",
                            "value": "docker-redis-cache:6379"
                        },
                        {
                            "name": "REGISTRY_AUTH",
                            "value": "token"
                        },
                        {
                            "name": "REGISTRY_AUTH_TOKEN_REALM",
                            "value": "[redacted]"
                        },
                        {
                            "name": "REGISTRY_AUTH_TOKEN_SERVICE",
                            "value": "[redacted]"
                        },
                        {
                            "name": "REGISTRY_AUTH_TOKEN_ISSUER",
                            "value": "[redacted]"
                        },
                        {
                            "name": "REGISTRY_AUTH_TOKEN_ROOTCERTBUNDLE",
                            "value": "/auth/auth.crt"
                        }
                    ],
                    "ports": [{
                        "containerPort": 5000,
                        "name": "registry",
                        "protocol": "TCP"
                    }],
                    "volumeMounts": [{
                        "name": "auth-signing",
                        "mountPath": "/auth",
                        "readOnly": true
                    }]
                }]
            }
        }
    }
}
```

Registry running with a 3rd-party S3 backend (minio).

---

Findings: `registry:2` breaks any layers larger than 100MB (`docker push` is fine, however `docker pull` gives error).

However, changing the image to `registry:2.5.1` or `registry:2.5`, and no more corruptions happened. Something must have changed between 2.5.x and 2.6.

**Comment by dmcgowan** [comment]
This appears to be a duplicate of https://github.com/docker/distribution/issues/2071

We would like more information on this if you can reach out to minio and figure it out. If you get an issue number please link it here. Also if there is a resolution we would like to be able to document a work around.

**Comment by zllovesuki** [comment]
@dmcgowan If minio's implementation is indeed to blame, then `registry:2.5.1` or `registry:2.5` will yield the same error. However, `filesystem layer verification failed` only appears on `registry:2`

**Comment by harshavardhana** [comment]
> @dmcgowan If minio's implementation is indeed to blame, then registry:2.5.1 or registry:2.5 will yield the same error

Can you open a bug @zllovesuki will see what needs to be fixed.

**Comment by zllovesuki** [comment]
@harshavardhana I meant that minio is not the problem, sorry if that wasn't clear to you.

It is possible that 63468ef4a85fabb756e799c2534f0df2f3c8167c changed the multipart behavior and minion couldn't handle it.

**Comment by dmcgowan** [comment]
@zllovesuki we'll investigate further. Does look similar to that issue with s3.

**Comment by harshavardhana** [comment]
> @harshavardhana I meant that minio is not the problem, sorry if that wasn't clear to you.

Okay understood - thanks @dmcgowan @zllovesuki

**Comment by dmcgowan** [comment]
@harshavardhana we can't find any reproductions using aws s3. The other person to experience this corruption was also on a non-minio 3rd party s3. Is it possible that minio has a bug with http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPartCopy.html being called with all the parts in parallel? That was added in 2.6 and the Docker Hub team has been running with this change on aws s3 since September without any reported corruption issues.

**Comment by harshavardhana** [comment]
> @harshavardhana we can't find any reproductions using aws s3. The other person to experience this corruption was also on a non-minio 3rd party s3. Is it possible that minio has a bug with http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPartCopy.html being called with all the parts in parallel? That was added in 2.6 and the Docker Hub team has been running with this change on aws s3 since September without any reported corruption issues.

We can try and see but with 2.6 i have been successfully using it locally as well. So not sure where the issue is coming from.  Also where does the corruption occur..  does the corruption occur in pulling the images or pushing ? 

Minio recently started supporting MultipartUploadCopyPart API in lieu of this change from registry 2.6 so it could be perfectly possible to have a bug. 

Let me take a look again..

**Comment by zllovesuki** [comment]
@harshavardhana It corrupts the image during `docker push` when using `registry:2` with minio version 2017-02-16T01:47:30Z.

**Comment by harshavardhana** [comment]
> @harshavardhana It corrupts the image during docker push when using registry:2 with minio version 2017-02-16T01:47:30Z.

Thanks will take a look. @zllovesuki @dmcgowan

**Comment by harshavardhana** [comment]
Was able to reproduce this after a successful push 

```
 docker pull localhost:5000/myoss
Using default tag: latest
latest: Pulling from myoss
9e1d18812a0d: Verifying Checksum 
f49ebe34919a: Download complete 
5fff08895e28: Verifying Checksum 
c933b00c3409: Download complete 
b2df646b60f9: Download complete 
ceeb7ccd5b89: Download complete 
d37006b2fcce: Downloading [==========================================>        ] 113.1 MB/131.6 MB
bd98d018ec49: Download complete 
fdcc34faa088: Verifying Checksum 
7e36e8b96543: Download complete 
31d5acf2c232: Download complete 
3a0c2fddab9a: Verifying Checksum 
2c338306b48f: Verifying Checksum 
filesystem layer verification failed for digest sha256:2c338306b48f0ebe2431b447a3aebed3741825cdc464d9c18aa5410497503ea9
```

**Comment by zllovesuki** [comment]
@harshavardhana Yes, that's exactly what happened. The images were corrupted during push. 

However, other non-S3 implementation has similar issues as well. Do you think this is specific to minio? Or this is somehow an implementation issue by registry.

**Comment by harshavardhana** [comment]
> @harshavardhana Yes, that's exactly what happened. The images were corrupted during push.
> 
> However, other non-S3 implementation has similar issues as well. Do you think this is specific to minio? Or this is somehow an implementation issue by registry.

Actually found the issue in minio the offsets for CopyObjectPart were wrongly interpreted on our end. It works fine with the fix. 

Will paste it here can give you a docker image if you can validate? @zllovesuki

**Comment by zllovesuki** [comment]
@harshavardhana Sure thing, I can validate :) Should I compile myself or you have a link for me?

**Comment by harshavardhana** [comment]
> @harshavardhana Sure thing, I can validate :) Should I compile myself or you have a link for me?

Sorry for the delay, Yes you can pull from  `docker pull y4m4/minio` and let us know. Thanks @zllovesuki

**Comment by zllovesuki** [comment]
@harshavardhana lol nevermind. I manually pull your repo and build it. I can verify that it does fix the corruption issue.

**Comment by harshavardhana** [comment]
> @harshavardhana lol nevermind. I manually pulled your repo and built it. I can verify that it does fix the corruption issue.

Great thanks! for letting us know. If you can :+1: on the PR will take it in - https://github.com/minio/minio/pull/3838

**Comment by dmcgowan** [comment]
@harshavardhana thanks for figuring this out, that is pretty awesome!

Once that PR is merged, mention it here and any necessary version, then we can close this issue out.

**Comment by zllovesuki** [comment]
@harshavardhana when can we expect a new release with this patched?

**Comment by harshavardhana** [comment]
> @harshavardhana when can we expect a new release with this patched?

Yes soon, there are some UI changes pending should be happen next week.  For now you can use `minio/minio:edge`

**Comment by xaqq** [comment]
Hello,
Just to confirm that using patched minio (edge) fixed the issue for me.

**Comment by dmcgowan** [comment]
Going to close this out, once it makes it into a release feel free to mention here.

Thanks!

**Comment by harshavardhana** [comment]
> Going to close this out, once it makes it into a release feel free to mention here.

We made a new Minio release - @dmcgowan @xaqq @zllovesuki
